feat: 🎸 Update hotkeys and user preferences modal

igoroctaviano opened the pull request:
This feature fix incompatibility with existent hotkeys component and
adds user preferences modal back

Closes: #923

ladeirarodolfo commented on platform/viewer/src/connectedComponents/ConnectedUserPreferencesModal.js:
```diff
@@ -0,0 +1,59 @@
+};
+
+/* HotkeysManager's hotkeysDefinitions have different output/input. */
+const hotkeysManagerFormatter = hotKeysData => {
```
Can we move this method to hotkeyManager file instead?

igoroctaviano replied:
Done! thx.

ladeirarodolfo commented on platform/viewer/src/connectedComponents/ConnectedUserPreferencesModal.js:
```diff
@@ -0,0 +1,59 @@
+      hotkeysManager.setHotkeys(
+        hotkeysManagerFormatter(cloneDeep(hotKeysData))
+      );
+      dispatch(setUserPreferences({ windowLevelData, hotKeysData }));
```
Considering you are closing modal anyway (assuming this is the right approach), can you invert the other? first you close then you dispatch to store (this kind of prevents user of dispatching twice). Same bellow

igoroctaviano replied:
Done! thx.

igoroctaviano commented:
> What if there is one tool active and we change the hotkey for it? did you test this scenario? I`m wondering if every related place will be updated properly

@ladeirarodolfo can you give more details about this execution flow? Since we are using the hotkey manager to update the definitions, it should guarantee consistency.

dannyrb commented:
This appears to be working well! Before I give it a code review, how hard would this be to update to use the just merged modal provider?

igoroctaviano commented:
> This appears to be working well! Before I give it a code review, how hard would this be to update to use the just merged modal provider?

Done!

dannyrb commented:
![image](https://user-images.githubusercontent.com/5797588/68314198-cb50b000-0083-11ea-99a2-0ae47f1f44bd.png)

dannyrb commented on platform/ui/src/components/userPreferencesForm/UserPreferencesForm.styl:
```diff
@@ -15,3 +15,14 @@
 .ModalHeader
```
Are we still using this style? Do we need to remove it to make sure it doesn't accidentally impact something else?

igoroctaviano replied:
This was fixed with the latest changes. thx.

igoroctaviano replied:
This was fixed with the latest changes. thx.

dannyrb commented on platform/ui/src/elements/form/DropdownMenu.css:
```diff
@@ -69,6 +69,7 @@
   color: var(--text-color-active);
   padding: 10px 15px;
   border-bottom: 1px solid #ccc;
+  width: 100%;
```
Are we using this anywhere else? Did we test to make sure this didn't negatively impact other usage?

igoroctaviano replied:
This was fixed with the latest changes. thx.

dannyrb commented on platform/viewer/cypress/integration/common/OHIFStudyViewer.spec.js:
```diff
@@ -267,7 +267,7 @@ describe('OHIF Study Viewer Page', function() {
 
   it('opens About modal and verify the displayed information', function() {
-    cy.get('[data-cy="options-menu"]').click();
+    cy.get('[data-cy="options-menu"]').click({ multiple: true });
```
I think this means we click once per item returned. Is that really what we want? We probably on the "first" item to be clicked.

igoroctaviano replied:
This was fixed with the latest changes. thx.

dannyrb commented on platform/viewer/src/connectedComponents/ConnectedHeader.js:
```diff
@@ -1,10 +1,18 @@
+    state.preferences && !isEmpty(state.preferences.hotKeysData)
+      ? state.preferences.hotKeysData
+      : hotkeysManager.hotkeyDefinitions;
+  hotkeysManager.setHotkeys(hotkeysManager.format(cloneDeep(newHotKeysData)));
```
Instead of formatting/cloning here, can we offload that concern to the hotkeys manager?

igoroctaviano replied:
This was fixed with the latest changes. thx.

dannyrb commented on platform/core/src/classes/HotkeysManager.js:
```diff
@@ -1,3 +1,4 @@
+import cloneDeep from 'lodash.clonedeep';
```
Because we're cloning something simple, and we know the shape, I wonder if we can do it without pulling in this dep and in a more performant way.

Not blocking. Just food for thought.

dannyrb commented on platform/ui/src/components/userPreferencesForm/HotKeysPreferences.js:
```diff
@@ -10,31 +10,14 @@ import PropTypes from 'prop-types';
-      })
-    ).isRequired,
-    onChange: PropTypes.func,
+    hotkeyDefinitions: PropTypes.array.isRequired,
```
Might be nice to be more explicit about this. `PropTypes.arrayOf(PropTypes.shape({}))`

dannyrb commented on platform/ui/src/components/userPreferencesForm/HotKeysPreferences.js:
```diff
@@ -10,31 +10,14 @@ import PropTypes from 'prop-types';
-
     this.state = {
-      hotKeys: localHotKeys,
+      hotKeys: this.props.hotkeyDefinitions,
```
I assume this is the "local" form state?

igoroctaviano replied:
Yes, this is the local form state.

dannyrb commented on platform/ui/src/components/userPreferencesForm/HotKeysPreferences.js:
```diff
@@ -89,7 +76,7 @@ export class HotKeysPreferences extends Component {
         keyDownEvent.key ||
         String.fromCharCode(keyDownEvent.keyCode);
-      pressedKeys.push(keyName.toUpperCase());
+      pressedKeys.push(keyName);
```
Do all browsers use lowercase? do we need to guard here?

dannyrb commented on platform/ui/src/components/userPreferencesForm/UserPreferences.js:
```diff
@@ -8,14 +8,14 @@ import './UserPreferences.styl';
   // TODO: Make this more generic. Tabs should not be restricted to these entries
   static propTypes = {
-    hotKeysData: PropTypes.object.isRequired,
+    hotkeyDefinitions: PropTypes.array.isRequired,
```
Similarly, we may want an `arrayOf` `shape` here

dannyrb commented:
`PR_CHECKS` is passing, but I get this in Netlify:

![image](https://user-images.githubusercontent.com/5797588/68635393-5ee01180-04c6-11ea-9cf2-4246f8dfda47.png)

dannyrb commented:
This works well enough for me. I'm going to merge this :+1:

igoroctaviano commented:
> `PR_CHECKS` is passing, but I get this in Netlify:
> 
> ![image](https://user-images.githubusercontent.com/5797588/68635393-5ee01180-04c6-11ea-9cf2-4246f8dfda47.png)
> 
> Doesn't persist in a "Private" tab. Could be app caching; or it blows up because we changed the hotkey schema and it's trying to restore from `localstate` or `sessionstate`?

It blows up because we changed the hotkey schema and its trying to restore, it happens locally as well.
